compiler/rustc_arena/src/lib.rs

```diff
@@ -191,17 +195,31 @@ impl<T> TypedArena<T> {
     #[inline]
     pub fn alloc_from_iter<I: IntoIterator<Item = T>>(&self, iter: I) -> &mut [T] {
         assert!(mem::size_of::<T>() != 0);
-        let mut vec: SmallVec<[_; 8]> = iter.into_iter().collect();
-        if vec.is_empty() {
+
+        let intermediate: &mut Vec<T> = unsafe { &mut *self.intermediate.get() };
```

Comment:
The next function of the iterator might call `alloc_from_iter` again. I'm not sure about the soundness of holding multiple mutable references of intermediate on the stack. Maybe it would be better to put this into the for loop. This instruction should be pretty cheap anyways, mostly it's just adding pointers, and if optimizers are smart they can cache this between each run through the loop.

Reply:
The next iteration may, indeed, but we only mutate the vector after the inner iterator finishes. The question is, does there exist a code transformation that is allowed to change this?

Reply:
> but we only mutate the vector after the inner iterator finishes.

The [UnsafeCell](https://doc.rust-lang.org/stable/std/cell/struct.UnsafeCell.html) docs say:

> A &mut T must always be unique.

And:

> There is *no* legal way to obtain aliasing &mut, not even with UnsafeCell<T>.

If you used `Cell` instead of `UnsafeCell` you'd get an unwrap error when the second level of `alloc_from_iter` tries to get a mutable reference. You could either call `as_ptr` instead, and work with `* mut`, or limit the lifetime of the `&mut` so that it doesn't stick around when the for loop's next function is called. It's a solvable problem, not the end of this PR, unless I'm missing something.

Reply:
I know about `Cell` and why this is iffy :) I guess you are right and taking hold of the reference inside the loop shouldn't be much of an issue and is actually the logically sound thing to do. It's just unergonomic :)